File: 2001~2500/2009. Minimum Number of Operations to Make Array Continuous/main.py
"""
Source: https://rb.gy/gylup
Date: 2023/7/27
Skill:
Ref:
Runtime: 1103 ms, faster than 25.44%
Memory Usage: 40.78 MB, less than 5.26%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def minOperations(self, nums: List[int]) -> int:
        sz, res, right, num_set, cnt = len(nums), sys.maxsize, 0, set(), defaultdict(int)
        nums.sort()
        for i in range(sz):
            while right < sz and nums[right] <= nums[i] + sz - 1:
                num_set.add(nums[right])
                cnt[nums[right]] += 1
                right += 1
            res = min(res, sz - len(num_set))
            cnt[nums[i]] -= 1
            if cnt[nums[i]] == 0:
                num_set.remove(nums[i])
        return res


if __name__ == "__main__":
    s = Solution()
    nums = [1, 2, 3, 5, 6]
    res = s.minOperations(nums)

File: 2001~2500/2009. Minimum Number of Operations to Make Array Continuous/main.cpp
/**
 * Source: ibit.ly/UbwFK
 * Date: 2023/10/10
 * Skill:
 * Runtime: 151 ms, faster than 73.01% of C++ online submissions
 * Memory Usage: 64.80 MB, less than 67.15% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    int minOperations(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int res = 1, sz = nums.size(), left = 0, num = 1;
        for (int i = 1; i < sz; ++i) {
            if (nums[i] != nums[i - 1])
                num++;
            while (nums[i] - nums[left] >= sz) {
                if (nums[left] != nums[left + 1])
                    num--;
                left++;
            }
            res = max(res, num);
        }
        return sz - res;
    }
};


int main() {
    Solution s;
    vector<int> nums{2, 3, 3, 2};
    vector<vector<int>> arr{{1,2}, {2,3}, {2,4}};
}